lib/shared/bucketing-assembly-script/assembly/types/dvcUserPB.ts

```diff
@@ -0,0 +1,100 @@
+import { DVCUser_PB } from './protobuf-generated/DVCUser_PB'
+import { NullableString } from './protobuf-generated/NullableString'
+import { NullableDouble } from './protobuf-generated/NullableDouble'
+import { CustomDataValue } from './protobuf-generated/CustomDataValue'
+import { _getPlatformData } from '../managers/platformDataManager'
+import { NullableCustomData } from './protobuf-generated/NullableCustomData'
+import { CustomDataType } from './protobuf-generated/CustomDataType'
+
+export class CustomDataValueInterpreter {
+    static isString(val: CustomDataValue): bool {
+        return val.type === CustomDataType.Str
+    }
+    static isFloat(val: CustomDataValue): bool {
+        return val.type === CustomDataType.Num
+    }
+    static isBool(val: CustomDataValue): bool {
+        return val.type === CustomDataType.Bool
+    }
+    static isNull(val: CustomDataValue): bool {
+        return val.type === CustomDataType.Null
+    }
+
+    static asString(val: CustomDataValue): string {
+        return val.stringValue
+    }
+
+    static asNumber(val: CustomDataValue): f64 {
+        return val.doubleValue
+    }
+
+    static asBool(val: CustomDataValue): bool {
+        return val.boolValue
+    }
+}
+
+export class DVCPopulatedUserPB {
+    readonly user_id: string
+    readonly email: string | null
+    readonly name: string | null
+    readonly language: string | null
+    readonly country: string | null
+    readonly appVersion: string | null
+    readonly appBuild: f64
+    customData: Map<string, CustomDataValue> | null
+    privateCustomData: Map<string, CustomDataValue> | null
+    readonly deviceModel: string | null
+
+    readonly createdDate: Date
+    readonly lastSeenDate: Date
+    readonly platform: string
+    readonly platformVersion: string
+    readonly sdkType: string
+    readonly sdkVersion: string
+    readonly hostname: string | null
+
+    constructor(user: DVCUser_PB) {
+        this.user_id = user.userId
+        this.email = deNullString(user.email)
+        this.name = deNullString(user.name)
+        this.language = deNullString(user.language)
```

Comment:
would this function call be pass by reference or by value for these strings? I wonder if calling a helper here instead of doing it inline actually results in an extra allocation?

Reply:
yeah it is I checked the WAT file